cli/medperf/comms/entity_resources/resources.py

```diff
@@ -1,135 +1,177 @@
+        expected_tarball_hash (str, optional): expected sha1 hash of tarball file
+        expected_folder_hash (str, optional): expected sha1 hash of uncompressed
+        version of the tarball file
 
     Returns:
-        str: Location where the additional_files.tar.gz file is stored locally.
+        tarball_hash (str): The hash of the downloaded file
+        folder_hash (str): The hash of the uncompressed version of the downloaded file
     """
-    add_path = config.additional_path
-    tball_file = config.tarball_filename
-    return get_cube_file(url, cube_path, add_path, tball_file)
+    additional_files_folder = os.path.join(cube_path, config.additional_path)
 
+    # If the additional_files folder exists, check if it has the expected hash
+    if expected_folder_hash and os.path.exists(additional_files_folder):
+        folder_hash = get_folder_sha1(additional_files_folder)
+        if folder_hash == expected_folder_hash:
+            return expected_tarball_hash, expected_folder_hash
 
-def get_benchmark_demo_dataset(
-    demo_data_url: str, uid: str = generate_tmp_uid()
-) -> str:
-    """Downloads the benchmark demo dataset and stores it in the user's machine
+    # Remove the folder if it exists
+    remove_path(additional_files_folder)
 
-    Args:
-        demo_data_url (str): location of demo data for download
-        uid (str): UID to use for storing the demo dataset. Defaults to generate_tmp_uid().
+    # Download the tarball file as usual
+    output_path = os.path.join(additional_files_folder, config.tarball_filename)
+    tarball_hash = download_resource(url, output_path, expected_tarball_hash)
+
+    # Untar and keep track of the folder hash.
+    untar(output_path)
+    folder_hash = get_folder_sha1(additional_files_folder)
+    return tarball_hash, folder_hash
 
-    Returns:
-        str: path where the downloaded demo dataset can be found
-    """
-    tmp_dir = storage_path(config.demo_data_storage)
-    demo_data_path = os.path.join(tmp_dir, uid)
-    tball_file = config.tarball_filename
-    filepath = os.path.join(demo_data_path, tball_file)
 
-    # Don't re-download if something already exists with same uid
-    if os.path.exists(filepath):
-        return filepath
+def get_benchmark_demo_dataset(url: str, expected_hash: str = None) -> str:
+    """Downloads and writes a demo dataset. If the hash is provided,
+    the downloaded file's integrity will be checked.
 
-    os.makedirs(demo_data_path, exist_ok=True)
+    Args:
+        url (str): URL where the compressed demo dataset file can be downloaded.
+        expected_hash (str, optional): expected sha1 hash of the downloaded file
 
-    download_resource(demo_data_url, filepath)
-    return filepath
+    Returns:
+        output_path (str): location where the compressed demo dataset file is stored locally.
+        hash_value (str): The hash of the downloaded file
+    """
+    # TODO: at some point maybe it is better to download demo datasets in
```

Comment:
My only con for this is if multiple benchmarks use the same demo dataset, in which case we would be wasting resources for redownloading already existing files. I wonder if this is a valid concern, given the probability of multiple benchmarks using the same dataset, and the assumption that demo datasets are small